### PR TITLE
Mixed Precision Transformer and Gradient Builder Refactor

### PR DESCRIPTION
**Pull request by iK1D**
Rebase previous PR on Cheng's changes. This PR includes:
- Refactor the mixed precision transformer to support other mixed precision data types other than FP16, only FP16 needs loss scale.
- Move mixed precision transformer before gradient builder, i.e., transform mixed precision on forward graph only.
- Add supporting mixed precision data types such as FP16/BFloat16 to gradient builder.
- Change Adam/Lamb optimizer to support more mixed precision data types other than FP16.

This PR is a preparation to support BFloat16 training. The target is to refactor the mixed precision transformer and gradient builder and make sure current FP32 and FP16 training works as usual. So it does NOT include any frontend parameter change or CUDA change to support new mixed precision data type (BFloat16).

Pref test result on Dev3:
FP16, seq=128
Max bs=66
Throughput: 190.402 (Adam)
Throughput: 183.76 (Lamb)

FP16, seq=512
Max bs=10
Throughput: 37.5534 (Adam)
Throughput: 36.1645 (Lamb)

FP32, seq=128
Max bs = 33
Throughput: 43.8391 (Adam)
Throughput: 42.9689 (Lamb)

FP32, seq=512
Max bs= 5
Throughput: 9.67252 (Adam)
Throughput: 9.43377 (Lamb)

Curve comparison for BERT:
image

Curve comparison for GPT2:
image

Curve comparison for UniLMv2:
image

### CODE REVIEW

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (weixingzhang) [inline orttraining/orttraining/core/graph/training_op_defs.cc:196]: Mixed precision training means "some ops are computed in float32, some other ops are computed in other precision, e.g. float16, bfloat16 for now, in the future, maybe int8". If we call it mixed precision data type, it seems confusing. Is it possible to find a better name?

File: orttraining/orttraining/models/runner/training_runner.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/models/runner/training_runner.cc:132]: >params_.allreduce_in_fp16; [](start = 48, length = 26)

need update?

File: orttraining/orttraining/models/runner/training_runner.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/models/runner/training_runner.cc:109]: >use_fp16_initializer [](start = 50, length = 20)

do we also need to update the param's name ?

File: orttraining/orttraining/core/graph/gradient_builder_base.h
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder_base.h:211]: I think this is not used anywhere now... maybe delete

File: orttraining/orttraining/core/graph/gradient_builder.cc
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc:100]: I had a discussion with Sherlock before on this. Since the implementation needs OneConstantNode, and when moving gradient builder after mixed precision, we need to pass in the element type as parameter. If we still use function body to do this, we need to add a new attribute to the TanhGrad schema for this purpose. Moving the implementation here seems more clean and clear. If in the future we need a kernel install of such sub-graph, we can change back.

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:196]: Any suggestion?

File: orttraining/orttraining/core/session/training_session.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/core/session/training_session.cc:451]: const pointer seems better here? since it could be empty sometimes...

File: orttraining/orttraining/core/graph/gradient_builder.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:139]: Matmul's inputs and output should share the same type. 

So all the type selection in this function can share the same type parameter...

**Comment by SherlockNoMad** [comment]
PR mostly looks good to me. 
Please remember to run E2E test and frontend E2E test. As the training graph is slightly modified, the expected loss value might need to be updated.

**Review by SherlockNoMad — DISMISSED** [review]

File: orttraining/orttraining/core/graph/gradient_builder.cc
Comment (weixingzhang) [reply orttraining/orttraining/core/graph/gradient_builder.cc:100]: The original design was decided to associate the function body of grad of an op with a subgraph which is composed by a few primitive ops. If there is no native grad for that op then the function body will be inlined.  The change of using function body was initially done by @souptc. So, I would leave it to him to comment.

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (weixingzhang) [reply orttraining/orttraining/core/graph/training_op_defs.cc:196]: no good suggestion.

File: orttraining/orttraining/core/graph/gradient_builder_base.h
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder_base.h:211]: GetSqrtGradient uses this one.

**Comment by iK1D** [comment]
> PR mostly looks good to me.
> Please remember to run E2E test and frontend E2E test. As the training graph is slightly modified, the expected loss value might need to be updated.

Ran the E2E test and got same loss value as master branch. Actually the training graph is same expect some grad arg_name.

**Comment by SherlockNoMad** [comment]
Hi Vincent, the PR looks good to me. 
But lets merge it after the rc3/ignite branch off, since this is not strictly needed for any production models. 

So let's merge this after Wednesday.

**Review by SherlockNoMad — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]

**Pull request merged**
